### ISSUE TITLE
Generate 2048 bit DHparam file to mitigate logjam attack

### ISSUE THREAD

**Issue by neuhaus** [al3x/sovereign#373]
See https://weakdh.org/sysadmin.html

We need to generate a 2048+ bit dhparam file (using `openssl dhparam`) and use it for Apache, OpenVPN and Postfix.
I have submitted a separate issue and pull request for Dovecot because it will generate its own dhparam file.

**Comment by neuhaus** [comment]
This could be done in the file  `roles/common/tasks/ssl.yml` I suppose

**Comment by neuhaus** [comment]
It will break Java 6 clients because they do not support DH parameters > 1024 bits.

**Comment by al3x** [comment]
`roles/common/tasks/ssl.yml` does seem the natural home for this task.

I don't think we need to support EOL'd software like Java 6.

**Comment by winks** [comment]
Prosody also supports a `dhparam` option in the `ssl = {}` section.

**Comment by neuhaus** [comment]
Solved for Postfix.

Apache is a bit more tricky:
Wheezy ships with Apache 2.2. Too old.
Trusty ships with Apache 2.4.7 and OpenSSL 1.0.1f. There, the dhparams can be appended to the certificates.

For Jessie and Xenial, starting with Apache 2.4.8 and OpenSSL 1.0.2, there is a new mechanism:

    SSLOpenSSLConfCmd DHParameters "{path to dhparams.pem}"

**Comment by ypid** [comment]
[debops.dhparam](https://github.com/debops/ansible-dhparam) might be worth checking out …

**Comment by neuhaus** [comment]
@ypid debops.dhparam looks good but it requires Ansible 1.9 and it doesn't help with the Apache 2.2/Apache 2.4 issue.

**Comment by mikeashley** [comment]
I investigated this on the jessie branch for Apache.  Debian 8 stable uses openssl 1.0.1k, so the DH parameters would need to be appended to the site certificates.

However, the current server configuration seems to already mitigate logjam.  I tested my server from [here](https://weakdh.org/sysadmin.html), and it reported my server as safe.  I also reviewed the report returned for my site by the [Qualys SSL Labs tester](https://www.ssllabs.com/ssltest/index.html).  It indicated that we're not using common DH primes, and we're not reusing DH public server params.

Does anything actually need to be done for Apache?  I'm not a security expert; just reporting what I'm seeing.  Could somebody more qualified assess this?

**Comment by mikeashley** [comment]
I think openvpn is ok.  The server configuration uses DH parameters.  The key size in the defaults is 2048 bits.  The task list generates the parameters.

Summary of where I think we're at: postfix and dovecot are ok, openvpn is ok, apache needs a look (see previous comment).

**Comment by mikeashley** [comment]
I looked at this again.  We are running Apache 2.4.10 on master.  I reviewed the [mod_ssl documentation](http://httpd.apache.org/docs/current/mod/mod_ssl.html).  It looks like we are "ok" with no changes, since mod_ssl uses standardized DH parameters corresponding to key length.  Logjam appears to be mitigated.

However, we may still want to use the custom DH parameters we are generating for postfix for some additional security as well as consistency.  It's easy enough to reference the parameters using the new [SSLOpenSSLConfCmd](http://httpd.apache.org/docs/current/mod/mod_ssl.html#sslopensslconfcmd) directive, but we would need to use OpenSSL 1.0.2 from Debian backports.

I'm recommending we do this and close this issue.  Any feedback?

**Comment by neuhaus** [comment]
Yes, we finally reached the stage described in [my comment from last December](https://github.com/sovereign/sovereign/issues/373#issuecomment-163578499)!

**Comment by mikeashley** [comment]
I am still troubleshooting why I can't get Apache to accept the `SSLOpenSSLConfCmd` directive.

Upgrading the libssl1.0.0 package from jessie-backports is not sufficient.  It does update libssl to 1.0.2h, but the next problem appears to be that `mod_ssl.so` in jessie does not have support for `SSLOpenSSLConfCmd` compiled into the module.  At least, `strings mod_ssl.so` does not give me any references to `SSLOpenSSLConfCmd`.  I am guessing its build was configured with a version of libssl < 1.0.2.  I can confirm this hypothesis by building apache 2.4.10 from source with a version of libssl >= 1.0.2 and then checking its `mod_ssl.so`.

I'm posting this in case somebody can tell me I'm on the wrong track or if somebody is using `SSLOpenSSLConfCmd` and can share their configuration with me.

**Comment by mikeashley** [comment]
Read the source code for mod_ssl in Apache 2.4.10.  Indeed, it elides support for the `SSLOpenSSLConfCMD` if compiled against an unsupporting version of libssl.  Skimming dependencies, I think we'll need to go after the Apache package from Debian testing.

**Comment by mikeashley** [comment]
Or wait for Apache 2.4.23 to land in backports, which is where I am leaning.  I am concerned about breaking dependencies if we pull from Debian testing.  Logjam is currently mitigated, and it doesn't seem worth the configuration headache to use custom DH parameters.

**Comment by neuhaus** [comment]
Is there an open issue already on debian regarding the Apache 2.4.10 issue?
What does Apache 2.4.23 improve in this regard?

**Comment by mikeashley** [comment]
Pulled the jessie tag since it's master now.
